tap_google_ads/streams.py

```diff
@@ -80,9 +80,60 @@ def build_parameters():
     param_str = ",".join(f"{k}={v}" for k, v in API_PARAMETERS.items())
     return f"PARAMETERS {param_str}"
 
+def generate_where_and_orderby_clause(last_pk_fetched, filter_param, composite_pks):
+    """
+    Generates a WHERE clause and a ORDER BY clause based on filter parameter(`key_properties`), and
+    `last_pk_fetched`.
+
+    Example:
+
+    Single PK Case:
+
+    filter_param = 'id'
+    last_pk_fetched = 1
+    composite_pks = False
+    Returns:
+    WHERE id > 1 ORDER BY id ASC
+
+    Composite PK Case:
+
+    composite_pks = True
+    filter_param = 'id'
+    last_pk_fetched = 1
+    Returns:
+    WHERE id >= 1 ORDER BY id ASC
+    """
+    where_clause = ""
+    order_by_clause = ""
+
+    # Even If the stream has a composite primary key, we are storing only a single pk value in the bookmark.
+    # So, there might be possible that records with the same single pk value exist with different pk value combinations.
+    # That's why for composite_pks we are using a greater than or equal operator.
+    comparison_operator = ">="
+
+    if not composite_pks:
+        # Exclude equality for the stream which do not have a composite primary key.
+        # Because in single pk case we are sure that no other record will have the same pk.
+        # So, we do not want to fetch the last record again.
+        comparison_operator = ">"
+
+    if filter_param:
+        # Create ORDER BY clause for the stream which support filter parameter.
+        order_by_clause = f"ORDER BY {filter_param} ASC"
+
+    if last_pk_fetched:
+        # Create WHERE clause based on last_pk_fetched.
+        where_clause = f'WHERE {filter_param} {comparison_operator} {last_pk_fetched} '
+
+    return f'{where_clause}{order_by_clause}'
```

Comment:
Super minor, but since this is an f string now we can just put the space between the two parts here and not need the space at the end of the string used above. Not having the space in the where_clause makes it less susceptible to bugs since each part doesn't need to know where it falls in the string it's being combined into.

Reply:
If `last_pk_fetched(interrupted state)` is available, then we are generating a query with a `WHERE` clause. So, here if we add space in the `return` statement between two parts and if the `where_clause ` is not available, then it will append extra space before `order_by_clause`. That's why added extra space at the end of `where_clause`.